### PR TITLE
Document the SessionDiagnostic derive macro.

### PR DESCRIPTION
**Pull request by jumbatm**
Documents the `#[derive(SessionDiagnostic)` macro implemented in https://github.com/rust-lang/rust/pull/75138.

cc @oli-obk

### CODE REVIEW

File: src/diagnostics.md [hunk 1]
```diff
@@ -338,6 +338,10 @@ if let Ok(snippet) = sess.source_map().span_to_snippet(sp) {
 err.emit();
 ```
 
+Alternatively, for less-complex diagnostics, the `SessionDiagnostic` derive
+macro can be used -- see [Creating Errors With SessionDiagnostic](./diagnostics/sessiondiagnostic.md).
```
Comment (oli-obk) [inline src/diagnostics.md:342]:
```suggestion
Alternatively, for less-complex diagnostics, the `SessionDiagnostic` derive macro
can be used -- see [Creating Errors With SessionDiagnostic](./diagnostics/sessiondiagnostic.md).
```

I think this line was too long (CI complained)

**Review by oli-obk — APPROVED** [review]
This is awesome! Thanks for writing it up. I think with this we can now start creating E-easy issues for various sites in the compiler where we could move to `SessionDiagnostic`

**Review by LeSeulArtichaut — DISMISSED** [review]
Some lines are too long and exceed the 100 characters limit required to pass CI. I'd suggest to use `[][]`-style links instead of the inline `[]()`-style links.

**Comment by LeSeulArtichaut** [comment]
Also this design looks fantastic! It's awesome that someone finally implemented it ❤️

File: src/diagnostics/sessiondiagnostic.md [hunk 2]
```diff
@@ -0,0 +1,97 @@
...
+// ...
+tcx.sess.emit_err(FieldAlreadyDeclared {
+    field_name: f.ident,
+    span: f.span,
+    prev_span,
+});
+```
+
+We see that using `SessionDiagnostic` is relatively straight forward.  The `#[error = "..."]`
+attribute is used to supply the error code for the diagnostic.  We are then annotate fields in the
```
Comment (camelid) [inline src/diagnostics/sessiondiagnostic.md:33]:
```suggestion
attribute is used to supply the error code for the diagnostic.  We then annotate fields in the
```

File: src/diagnostics/sessiondiagnostic.md [hunk 3]
```diff
@@ -0,0 +1,97 @@
...
+struct with various information of how to convert an instance of the struct into a proper
```
Comment (camelid) [inline src/diagnostics/sessiondiagnostic.md:34]: Maybe this?

```suggestion
struct with various information on how to convert an instance of the struct into a rendered
```

File: src/diagnostics/sessiondiagnostic.md [hunk 4]
```diff
@@ -0,0 +1,97 @@
...
+        diag.set_primary_message(format!("field `{field_name}` is already declared", field_name = self.field_name));
+        diag.span_label(self.span, "field already declared");
+        diag.span_label(self.prev_span, format!("`{field_name}` first declared here", field_name = self.field_name));
+        diag
+    }
+}
+```
+
+The generated code draws attention to a number of features.  First, we see that within the strings
+passed to each attribute, we see that field names can be referenced without needing to be passed
```
Comment (camelid) [inline src/diagnostics/sessiondiagnostic.md:52]: You already said “we see that”:

```suggestion
passed to each attribute, field names can be referenced without needing to be passed
```

**Review by camelid — COMMENTED** [review]
This feature looks pretty cool!

**Comment by jumbatm** [comment]
Thanks for the suggestions, all :)

@rustbot modify labels: waiting-on-review

**Review by mark-i-m — APPROVED** [review]
Thanks!

**Pull request merged**